zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneService.java

```diff
@@ -88,13 +171,305 @@ private ConnectionDetails buildConnectionDetails(String zoneName, IdentityZone i
     private String getZoneUri(String subdomain, String uaaUri) {
         URI uaaUriObject = URI.create(uaaUri);
         String currentUAAHostName = uaaUriObject.getHost();
+        String replacement = subdomain + "." + currentUAAHostName;
+        if (StringUtils.isEmpty(subdomain)) {
+            replacement = currentUAAHostName;
+        };
         URI newUAARoute = URI
-            .create(uaaUriObject.toString().replace(currentUAAHostName, subdomain + "." + currentUAAHostName));
+            .create(uaaUriObject.toString().replace(currentUAAHostName, replacement));
         return newUAARoute.toString();
     }
 
     @Override
     public void setApplicationEventPublisher(ApplicationEventPublisher applicationEventPublisher) {
         this.publisher = applicationEventPublisher;
     }
+
+    private String getSubDomainStr(IdentityZone identityZone) {
+        String id = identityZone.getId();
+        String subDomain = identityZone.getSubdomain();
+        if(id.equals(subDomain)){
+            subDomain = null;
+        }
+        return subDomain;
+    }
+
+    public void createZone(OrchestratorZoneRequest zoneRequest) throws OrchestratorZoneServiceException,
+                                                                                   ZoneAlreadyExistsException,
+                                                                                   AccessDeniedException {
+        if (!IdentityZoneHolder.isUaa()) {
+            throw new AccessDeniedException("Zones can only be created by being authenticated in the default zone.");
+        }
+        String name = zoneRequest.getName();
+        String adminClientSecret = getAdminClientSecret(zoneRequest);
+
+        checkOrchestratorZoneExists(name);
+
+        String subdomain = zoneRequest.getParameters().getSubdomain();
+        String id = UUID.randomUUID().toString();
+        subdomain = getSubDomain(subdomain, id);
+
+        String zoneSigningKey = createSigningKey(subdomain);
+
+        IdentityZone identityZone = geIdentityZone(subdomain, name, id, zoneSigningKey);
+
+        IdentityZone created = createIdentityZone(identityZone);
```

Comment:
The following statements are available in the existing controller, make sure it is available here also.

IdentityZoneHolder.set(created);
in finally - IdentityZoneHolder.set(previous);

Reply:
Done. Thanks!